src/Drupal10/Rector/Deprecation/AnnotationToAttributeRector.php

```diff
@@ -0,0 +1,282 @@
+
+use Drupal\Core\Action\Plugin\Action\Derivative\EntityPublishedActionDeriver;
+use Drupal\Core\Action\Attribute\Action;
+use Drupal\Core\Session\AccountInterface;
+use Drupal\Core\StringTranslation\TranslatableMarkup;
+
+/**
+ * Publishes an entity.
+ */
+#[Action(
+  id: 'entity:publish_action',
+  action_label: new TranslatableMarkup('Publish'),
+  deriver: EntityPublishedActionDeriver::class
+)]
+class PublishAction extends EntityActionBase {
+CODE_SAMPLE
+        )]);
+    }
+
+    /**
+     * @return array<class-string<Node>>
+     */
+    public function getNodeTypes(): array
+    {
+        return [Class_::class];
+    }
+
+    public function provideMinPhpVersion(): int
+    {
+        return PhpVersion::PHP_81;
+    }
+
+    /**
+     * @param Class_|ClassMethod $node
+     */
+    public function refactor(Node $node): ?Node
+    {
+        foreach ($this->configuration as $configuration) {
+            if ($this->rectorShouldApplyToDrupalVersion($configuration) === false) {
+                continue;
+            }
+
+            $result = $this->refactorWithConfiguration($node, $configuration);
+
+            // Skip if no result.
+            if ($result === null) {
+                continue;
+            }
+
+            return $result;
+        }
+
+        return null;
+    }
+
+    /**
+     * @param Class_|ClassMethod                 $node
+     * @param AnnotationToAttributeConfiguration $configuration
+     */
+    public function refactorWithConfiguration(Node $node, VersionedConfigurationInterface $configuration): ?Node
+    {
+        $phpDocInfo = $this->phpDocInfoFactory->createFromNode($node);
+        if (!$phpDocInfo instanceof PhpDocInfo) {
+            return null;
+        }
+
+        $tagsByName = $phpDocInfo->getTagsByName($configuration->getAnnotation());
+        if ($tagsByName === []) {
+            return null;
+        }
+
+        $hasAttribute = false;
+        foreach ($node->attrGroups as $attrGroup) {
+            foreach ($attrGroup->attrs as $attr) {
+                if ($attr->name->toString() === $configuration->getAttributeClass()) {
+                    $hasAttribute = true;
+                    break 2;
+                }
+            }
+        }
+
+        $docBlockHasChanged = \false;
+        foreach ($tagsByName as $valueNode) {
+            if (!$valueNode->value instanceof GenericTagValueNode) {
+                continue;
+            }
+
+            if ($hasAttribute === false) {
+                $stringValue = $valueNode->value->value;
+                $stringValue = '{'.trim($stringValue, '()').'}';
+                $tokenIterator = $this->tokenIteratorFactory->create($stringValue);
+                $data = $this->arrayParser->parseCurlyArray($tokenIterator, $node);
+                $attribute = $this->createAttribute($configuration->getAttributeClass(), $data);
+                $node->attrGroups[] = new AttributeGroup([$attribute]);
+            }
```

Comment:
Can this do a better check that the annotation matching the ending class name?

Wouldn't this cause `@foo` to become `#Foo` attribute because its a generic string?

Reply:
Do you mean here; you get to this code after checking a tag is present as configured. Then we move on to check if there is already an attribute. If that is present, then we loopt through the tagnames and if there was no attribute matching the relevant attribute class from the configuration we start building the attribute based on the class passed into the rector by the configuration. 

Don't really see how you could break this, unless user error using their own config?

Reply:
More so just:

```
$blah = explode('\', $configuration->getAttributeClass());
$className = end($blah);
if ($blah === substr($stringValue, 1)) {
```

Not entirely configured. Just that the tag matches the attribute's class name. I'm going to bet there is a "malformed" phpdoc somewhere with what the phpdoc parser considers generic tags